pilot/pkg/networking/grpcgen/cds.go

```diff
@@ -0,0 +1,270 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpcgen
+
+import (
+	"fmt"
+
+	cluster "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
+	core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	tls "github.com/envoyproxy/go-control-plane/envoy/extensions/transport_sockets/tls/v3"
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+
+	networking "istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking/core/v1alpha3"
+	"istio.io/istio/pilot/pkg/networking/util"
+	"istio.io/istio/pilot/pkg/util/sets"
+	"istio.io/istio/pkg/config/host"
+)
+
+// BuildClusters handles a gRPC CDS request, used with the 'ApiListener' style of requests.
+// The main difference is that the request includes Resources to filter.
+func (g *GrpcConfigGenerator) BuildClusters(node *model.Proxy, push *model.PushContext, names []string) model.Resources {
+	filter := newClusterFilter(names)
+	clusters := make([]*cluster.Cluster, 0, len(names))
+	for defaultClusterName, subsetFilter := range filter {
+		builder, err := newClusterBuilder(node, push, defaultClusterName, subsetFilter)
+		if err != nil {
+			log.Warn(err)
+			continue
+		}
+		clusters = append(clusters, builder.build()...)
+	}
+
+	resp := make(model.Resources, 0, len(clusters))
+	for _, c := range clusters {
+		resp = append(resp, &discovery.Resource{
+			Name:     c.Name,
+			Resource: util.MessageToAny(c),
+		})
+	}
+	return resp
+}
+
+// newClusterFilter builds a filtering map to determine which clusters need to be built.
+// gRPC will usually request each subset individually, regardless of if a previous response included it.
+func newClusterFilter(names []string) map[string]sets.Set {
```

Comment:
nit: Can you comment that this is basically a map of "resource name without subset" -> "resource names with subsets"?